### PR TITLE
flatten dict for wandb

### PR DESCRIPTION
**Pull request by mosheman5**
Flatten the args before logging them to wandb.
PR by @AsafMazuz1

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -109,12 +107,14 @@ def modeling(
...
+# TODO check how to use hydra without path override
+@hydra.main(config_name="runs/main", config_path="conf")
 def main(args):
 
     # Set logger
-    _logger = wandb if not args.experiment.debug else Mock()
-    _logger.init(project="finding_willy", name=args.experiment.name)
+    # if not args.experiment.debug else Mock()
+    _logger = wandb
+    _logger.init(project="testing", name=args.experiment.name)
```
Comment (mosheman5) [inline src/train.py:117]: rename the project to 'finding_willy'

File: src/train.py [hunk 2]
```diff
@@ -138,6 +138,7 @@ def main(args):
 
     # Logging
     logger = Logger(_logger)
+    args = flatten(args)
```
Comment (mosheman5) [inline src/train.py:141]: Wouldn't override `args` since we need it in other parts in the code. Would rename and create new variable

File: src/utils.py [hunk 3]
```diff
@@ -13,6 +13,13 @@
 import pandas as pd
 import re
 
+import collections
+
+try:
+    collectionsAbc = collections.abc
+except AttributeError:
+    collectionsAbc = collections
```
Comment (mosheman5) [inline src/utils.py:21]: what's the use case for the try n except here?

File: src/utils.py [hunk 4]
```diff
@@ -440,3 +473,14 @@ def merge_with_checkpoint(run_args, checkpoint_args):
     run_args.data.sample_rate = checkpoint_args.data.sample_rate
     OmegaConf.set_struct(run_args, True)
     return run_args
+
+
+def flatten(d, parent_key='', sep='_'):
```
Comment (mosheman5) [inline src/utils.py:478]: please use `sep='.'` instead of `sep='_'`

File: src/utils.py [hunk 5]
```diff
@@ -102,17 +111,18 @@ def __init__(self,
         self.pred_list = []
         self.pred_proba_list = []
         self.label_list = []
-        self.metrics_dict = {'accuracy': [], 'f1score': [], 'precision': [], 'recall': [], 'auc': []}
+        self.metrics_dict = {'accuracy': [], 'f1score': [],
+                             'precision': [], 'recall': [], 'auc': []}
```
Comment (mosheman5) [inline src/utils.py:115]: Can you change your max line length in the code editor? it create a lot of diffs and it's hard to follow..

File: src/utils.py (hunk 3 above)
Comment (AsafMazuz1) [reply src/utils.py:21]: Fallback version for this library.
Depends on the python version

File: src/utils.py (hunk 5 above)
Comment (AsafMazuz1) [reply src/utils.py:115]: Set the value to?

Comment (Z30G0D) [reply src/utils.py:115]: 120 characters is default seems to be sufficient since most of us use IntelliJ's Pycharm

File: src/utils.py [hunk 6]
```diff
@@ -440,3 +473,14 @@ def merge_with_checkpoint(run_args, checkpoint_args):
     run_args.data.sample_rate = checkpoint_args.data.sample_rate
     OmegaConf.set_struct(run_args, True)
     return run_args
+
+
+def flatten(d, parent_key='', sep='.'):
```
Comment (Z30G0D) [inline src/utils.py:478]: Do we also need a reverse method (unflatten) to turn back to yaml?
Just thinking out loud

**Review by Z30G0D — COMMENTED** [review]
Very nice job!
Small comment.

File: src/train.py [hunk 7]
```diff
@@ -109,11 +107,13 @@ def modeling(
...
 
-@hydra.main(config_name="runs/main", config_path="conf")  # TODO check how to use hydra without path override
+# TODO check how to use hydra without path override
+@hydra.main(config_name="runs/main", config_path="conf")
 def main(args):
 
     # Set logger
-    _logger = wandb if not args.experiment.debug else Mock()
+    # if not args.experiment.debug else Mock()
+    _logger = wandb
```
Comment (noamzbr) [inline src/train.py:116]: why remove that and log debug runs?

File: src/utils.py [hunk 8]
```diff
@@ -440,3 +473,14 @@ def merge_with_checkpoint(run_args, checkpoint_args):
...
+    items = []
+    for k, v in d.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if isinstance(v, collectionsAbc.MutableMapping):
+            items.extend(flatten(v, new_key, sep=sep).items())
```
Comment (noamzbr) [inline src/utils.py:483]: Think `items.update(flatten(v, new_key, sep=sep))` is more readable but really not important

File: src/utils.py (hunk 6 above)
Comment (noamzbr) [reply src/utils.py:478]: I tend to agree, useful for actually running with past configurations. 
Among other things it would mean that we can't use . as a default separator. Note sure what we can use because nearly other char has a use in python (except , which is not elegant). Ideas?

**Review by noamzbr — CHANGES_REQUESTED** [review]
All in all looks simple and would do the work. I think an inverse operation is must (as @Z30G0D said) and that it would be better to remove the diffs originating in IDE settings on line length.

File: src/train.py (hunk 7 above)
Comment (AsafMazuz1) [reply src/train.py:116]: This was for my testing.
I will fix this in my next push :)

**Review by noamzbr — APPROVED** [review]

File: src/utils.py (hunk 6 above)
Comment (noamzbr) [reply src/utils.py:478]: @mosheman5 convinced me this is not needed
What we do need is a more fool proof approach to uploading experiments to s3
